### PR TITLE
refactor (webapi): authentication

### LINKED ISSUE #171: Add JSON Web Token Support

### PR DESCRIPTION
**Pull request by subotic**
### Summary

- Added JSON Web Token support
- For `v1` to the outside, nothing should have changed. Authentication is still session/cookie based, only now the session id is a JWT.
- For `v2` there are no session/cookies anymore. At `login` time, a token is returned, which then needs to be sent with every request either as an URL parameter or in the authorization header with the `Bearer` scheme.
- User profiles are now cached by the `UsersResponder`
- Login into `v1` and accessing `v2` resources works.
- Login into `v2` and accessing `v1` resources works.
- Update documentation.

### Issues

- Resolves #171

### CODE REVIEW

**Comment by subotic** [comment]
working on #171

File: docs/rst/knora-api-server/design-documentation/authentication.rst [hunk 1]
```diff
@@ -26,146 +26,139 @@ Authentication in the Knora API Server
...
 
-Authentication is the process of making sure that if someone is
-accessing something then this someone is actually also the someone
-he pretends to be. The process of making sure that someone is
-authorized, i.e. has the permission to access something, is handled as
-described in the section on authorization in the Knora base ontology
-document. TODO: add a link to this.
+Authentication is the process of making sure that if someone is accessing something then this someone is actually also
+the someone he pretends to be. The process of making sure that someone is authorized, i.e. has the permission to access
+something, is handled as described in the section on authorization in the Knora base ontology document. TODO: add a link to this.
```
Comment (benjamingeer) [inline docs/rst/knora-api-server/design-documentation/authentication.rst]: Could you add that link?

File: docs/rst/knora-api-server/design-documentation/authentication.rst [hunk 2]
```diff
@@ -26,146 +26,139 @@ Authentication in the Knora API Server
...
----------------
+--------------
 
-The authentication in Knora is based on Basic Auth
-`HTTP basic authentication`_, URL parameters, and cookies. This means that
-on every request (to any of the routes), an authentication header, URL
-parameters or cookie need to be sent.
+The authentication in Knora is based on Basic Auth `HTTP basic authentication`_, URL parameters, `JSON Web Token`_ and
+cookies. This means that on every request (to any of the routes), credentials need to be sent either via authorization
+header, URL parameters or cookie header.
```
Comment (benjamingeer) [inline docs/rst/knora-api-server/design-documentation/authentication.rst]: Does this mean that a JSON web token can be sent in an authorization header, in a URL parameter or in a cookie header?

File: docs/rst/knora-api-server/design-documentation/authentication.rst [hunk 3]
```diff
@@ -26,146 +26,139 @@ Authentication in the Knora API Server
...
-      "status": 2,
-      "message": "bad credentials: user found, but password did not match"
-    }
+Login and Logout
+^^^^^^^^^^^^^^^^
 
-**No credentials:**
+A client sends a POST request (``{"email":"usersemail", "password":"userspassword"}``) to the
+**/v2/authentication** route with *email* and *password* in the body. If the credentials are valid,
+a `JSON WEB Token`_ (JWT) will be sent back in the response. On all subsequent calls to any route,
```
Comment (benjamingeer) [inline docs/rst/knora-api-server/design-documentation/authentication.rst]: Could you add an example of what a response containing a JWT looks like?

File: docs/rst/knora-api-server/design-documentation/authentication.rst [hunk 4]
```diff
@@ -26,146 +26,139 @@ Authentication in the Knora API Server
...
-filter the result based on the users permissions.
+-------------------
+
+For authentication to work with the media server, we need to add support for cookies. At the moment the SALSAH-App
+would set BasicAuth heathers, but this only works for AJAX requests using ``SALSAH.ApiGet`` (``Put``, etc.).
+Since the medias are embedded as source tags, the browser would get them on his own, and doesn't know anything about
+the needed AuthHeathers. With cookies, the browser would send those automatically with every request. The media server
+can use the credentials of the user requesting something for accessing the RepresentationsRouteV1, i.e. make this
+request in the name of the user so to speak, then the RepresentationResponderV1 should have all the information it
+needs to filter the result based on the users permissions.
```
Comment (benjamingeer) [inline docs/rst/knora-api-server/design-documentation/authentication.rst]: This is what Sipi does already, right?

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala [hunk 5]
```diff
@@ -462,39 +600,117 @@ object Authenticator {
...
+    }
+}
+
+object JWTHelper {
+
+    import Authenticator.AUTHENTICATION_INVALIDATION_CACHE_NAME
+
+    val algorithm = Algorithm.HS256
+    val requiredHeaders: Set[HeaderField] = Set[HeaderField](Typ)
+    val requiredClaims: Set[ClaimField] = Set[ClaimField](Iss, Sub, Aud, Iat, Exp, Jti)
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: Please add comments explaining what is being configured here.

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala [hunk 6]
```diff
@@ -462,39 +600,117 @@ object Authenticator {
...
+      * @param secretKey the secret key used for encoding.
+      * @param longevity the token's longevity in days.
+      * @return a [[String]] containg the JWT.
+      */
+    def createToken(userIri: IRI, secretKey: String, longevity: Long): String = {
+
+        val headers = Seq[HeaderValue](Typ("JWT"), Alg(algorithm))
+
+        val now: Long = System.currentTimeMillis() / 1000l
+        val nowPlusLongevity: Long = now + longevity * 60 * 60 * 24
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: Please add a comment saying how much time this is.

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala [hunk 7]
```diff
@@ -462,39 +600,117 @@ object Authenticator {
...
+
+        val identifier: String = UUID.randomUUID().toString()
+
+        val claims = Seq[ClaimValue](Iss("webapi"), Sub(userIri), Aud("webapi"), Iat(now), Exp(nowPlusLongevity), Jti(identifier))
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: Please add comments explaining what claims are being made.

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala [hunk 8]
```diff
@@ -462,39 +600,117 @@ object Authenticator {
...
+      */
+    def validateToken(token: String, secret: String): Boolean = {
+
+        if (CacheUtil.get[UserProfileV1](AUTHENTICATION_INVALIDATION_CACHE_NAME, token).nonEmpty) {
+            // token invalidated so no need to decode
+            log.debug("validateToken - token found in invalidation cache, so not valid")
+            false
         } else {
-            throw BadCredentialsException(BAD_CRED_USERNAME_NOT_SUPPLIED)
+            DecodedJwt.validateEncodedJwt(
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: Please used named parameters in method calls with more than one parameter, it's less error-prone and is easier to read.

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala [hunk 9]
```diff
@@ -462,39 +600,117 @@ object Authenticator {
...
+    /**
+      * Extract the encoded user IRI.
+      *
+      * @param token  the JWT.
+      * @param secret the secret used to encode the token.
+      * @return an optional [[IRI]].
+      */
+    def extractUserIriFromToken(token: String, secret: String): Option[IRI] = {
+
+        DecodedJwt.validateEncodedJwt(
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: named parameters please

File: docs/rst/knora-api-server/design-documentation/authentication.rst (hunk 2 above)
Comment (subotic) [reply docs/rst/knora-api-server/design-documentation/authentication.rst]: Yes.

File: docs/rst/knora-api-server/design-documentation/authentication.rst (hunk 3 above)
Comment (subotic) [reply docs/rst/knora-api-server/design-documentation/authentication.rst]: added.

File: docs/rst/knora-api-server/design-documentation/authentication.rst (hunk 4 above)
Comment (subotic) [reply docs/rst/knora-api-server/design-documentation/authentication.rst]: Yes, more or less. It's a bit out of date.

File: docs/rst/knora-api-server/design-documentation/authentication.rst (hunk 1 above)
Comment (subotic) [reply docs/rst/knora-api-server/design-documentation/authentication.rst]: added

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala (hunk 5 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: done

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala (hunk 6 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: it's user configurable in `application.conf`:

```
app {
     ...
     // the JSON Web Token longevity in days.
     jwt-longevity = 30
     ...
}

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala (hunk 7 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: done

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala (hunk 8 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: done

File: webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala (hunk 9 above)
Comment (subotic) [reply webapi/src/main/scala/org/knora/webapi/routing/Authenticator.scala]: done

**Comment by benjamingeer** [comment]
Did you forget to push your commit? :)

**Comment by subotic** [comment]
Sorry, I'm answering as I go along. It's not yet pushed.

**Comment by subotic** [comment]
now it is pushed. thanks.

**Comment by tobiasschweizer** [comment]
I think you forgot to deactivate the debugging messages. Could you do that, please? Thanks!

**Comment by subotic** [comment]
yes, I have deactivated them in the `wip/beol_date` branch.

**Pull request merged**